design/20220722-publicbundle.md

```diff
@@ -0,0 +1,289 @@
+his clusters.
+
+He has to update the base image of every container the organisation builds to include the new root, but
+that task requires several steps:
+
+1. Ensuring he has a list of every base image used across the organization. If he misses a container,
+   it'll be vulnerable to MitM attacks until it's updated.
+2. Making sure that each distinct base image used across the organization (Alpine / Debian / Distroless / etc)
+   has actually received the updated trust store. There's no point updating if there's nothing to update to!
+3. Rebuilding every single container which has an embedded trust bundle across the organization. In practice,
+   this is nearly all of them.
+4. Testing the newly built containers in CI to ensure that the updated base images haven't changed any other
+   functionality. If some containers were running an out-of-support base image, they won't get the updated
+   ca-certificates package, and Charles might need to bump to a new minor version which could be incompatible
+   with his applications.
+5. After rebuilding, he has to roll all deployments of all pods to make sure they're using the new images.
+
+All of the above would have to happen under extreme pressure.
+
+### Risks and Mitigations
+
+#### Bundle Container Updates
+
+This proposal makes it absolutely critical that we have a smooth and rapid release process for the
+"bundle" container. A security incident involving a root certificate would be front-page news and a
+major incident for basically every organisation in the world and we'd have to treat it the same,
+to the point of waking engineers up in the middle of the night to cut a new release if required.
+
+#### Bundle Container Attack Surface
+
+Since this proposal includes a need for some kind of process by which the controller container can
+query the bundle from the bundle container, that necessitates an increase in the attack surface of
+trust as an overall application.
+
+The communication channel between the controller and bundle containers is absolutely security critical;
+if a malicious actor were able to intercept and modify the bundle in-flight it would allow them to insert
+their own root certificate and MitM every connection between any container using the resulting bundle.
+
+This is mitigated by the use of two containers in a Pod - the Kubernetes security model should allow
+for this to be done securely without interference.
+
+## Design Details
+
+### Bundle Container "Server"
+
+The initial version of the bundle container would simply embed the trust bundle into a small Go
+application, similar to what [this PR](https://github.com/cert-manager/trust/pull/36) does.
+
+This Go application would then expose the certificate to the controller pod by writing the bundle,
+to a known location on [a shared volume](https://kubernetes.io/docs/tasks/access-application-cluster/communicate-containers-same-pod-shared-volume/).
```

Comment:
With unix socket- I wonder if there are some extra authentication checks we could do with things like PID numbers.. thinking out load :slightly_smiling_face:

Reply:
> I would rather the two containers talk over a unix socket. This prevents the middle-man file which can be of course hijacked, but also means we can have instance bi-directional communication if either of them need a watch style stream.

How would you share the unix socket without making the socket itself a "middle-man file"? Would a unix socket not normally be presented as a file, itself?

> means we can have instance bi-directional communication if either of them need a watch style stream.

I think we should absolutely rule out any bi-directional communication.

That significantly complicates the design of both the bundle and the controller container, and therefore makes it _much_ harder to ensure that any version of the bundle container would work with any version of the controller.

In addition, it adds a tonne of failure modes and a lot of attack surface for the bundle container since it needs to be able to listen, parse input and respond.

Currently in this design the bundle is simply writing a JSON file to disk, which is trivial to reason about and trivial for the controller to consume. If the controller wants to watch that file for updates that's easy too. Bidirectional systems complicate things a lot, and I'm not sure what we gain from that.

Reply:
> How would you share the unix socket without making the socket itself a "middle-man file"? Would a unix socket not normally be presented as a file, itself?

In this context I mean there is no way to write to the file. The trust manager can be sure it’s only talking to one PID.

—-

Bidirectional communication would be required if we want to have “dynamic” side car stores. FS notify is no more or less trivial than listening on a unix socket.

I’m not necessarily is favour of having bidirectionalness either, but the design talks about the source being dynamic, so that was where my head was going.

—-

Regardless, I do think using a unix socket is going to be far less flaky and more reliable to make changes immediately. I also believe it to be less fallible than files on disk since socket connections objectively have more information about the client/server. 

The CSI mechanism in kube is some prior art here which is in the same domain and also uses unix sockets.

I don’t think the API will be any different for either approach.

Reply:
> Bidirectional communication would be required if we want to have “dynamic” side car stores. FS notify is no more or less trivial than listening on a unix socket.

Bidirectional comms explicitly _wouldn't_ be required under the proposed design; there's absolutely no need for the controller container to send any message or data to the bundle container under any circumstances. The controller container is a consumer of whatever file is written by the bundle container.

> I’m not necessarily is favour of having bidirectionalness either, but the design talks about the source being dynamic, so that was where my head was going.

By a dynamic source I mean that the bundle container would write the public bundle with updates after the first time; there's no need for bidirectional comms there since the controller can either use fsnotify or just read the file every X seconds. I think I could make this clearer in the design, which I'll do.

> In this context I mean there is no way to write to the file. The trust manager can be sure it’s only talking to one PID.

As far as I understand, it should be possible with a shared volume to have the bundle container run with (say) UID 2000 and to have the controller container run with (say) UID 1000.  If the bundle container writes the bundle with permissions `0o664`, then anyone can read but only the bundle container would be able to write.

Do you have any links which show how you'd expect the unix socket to be shared between the two containers?

Reply:
As discussed in the standup, here is a very toy example of the PID stuff I was talking about. You can see the general behavior by running `go run .` under `/controller` and `/sidecar` and re-running the sidecar program. 


[sandbox.zip](https://github.com/cert-manager/trust/files/9241127/sandbox.zip)

Reply:
That pretty much matches what I expected it to look like, yeah; I've used unix sockets like that before, locally.

```go
// sidecar/main.go
conn, err := net.Dial("unix", "/tmp/trust.sock")

// controller/main.go
listener, err := net.Listen("unix", "/tmp/trust.sock")            
```

The problem is how we share `trust.sock` between two containers in a pod. This design proposes using a shared volume to share the bundle JSON file. If we replace that JSON file with a unix socket called `trust.sock`, then we'd presumably use a shared volume to contain that socket file.

If we're replacing the JSON with a socket file, we _must_ have **at least** the same security concerns as we'd have with just the JSON blob. If we can't secure the JSON blob we can't secure the socket, because they're both just files.

Is there some other way to share the file between the containers?

I think I'm also missing why we'd actually want to assert the same PID using `SO_PEERCRED`. Apart from PID 1 (which isn't relevant here) I don't know of a reason to treat one PID differently to another, and I don't see any value in asserting that the same PID is used. Why would the controller care how the sidecar operates regarding its PIDs? What threat model are we using where we require protection against the sidecar changing its PID?

(I do see a reason for asserting the same UID is used, but we can let the filesystem do that for us by marking the file `0o664` or `0o644`)

Reply:
Ugggh this is all kinda moot :sob:  k8s `Deployment`s only allow a pod restartPolicy of `Always` so you can't have a container which exits with a success code in a Pod which is part of a Deployment; they _have_ to stay running and "hang" forever even if there's no more work to be done.

Given that the static bundle won't ever update in the lifecycle of a pod, I'd much prefer to not take up any resources after our work is done... but I guess Kubernetes will _force_ us to consume more resources here. At that point, a unix socket which just always returns the same content makes sense, and would at least allow the client to be simpler

Reply:
Actually I think this works reasonably well with minimal resource usage and would respect the pod being shut down by k8s

```go
package main

import (
	"os"
	"os/signal"
	"syscall"
)

func main() {
	// write file

	sigs := make(chan os.Signal, 1)

	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
	<-sigs
}
```